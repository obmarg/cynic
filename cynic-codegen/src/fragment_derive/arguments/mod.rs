mod analyse;
mod output;
mod parsing;

use proc_macro2::Span;

use crate::{
    error::Errors,
    schema::{Schema, Unvalidated},
};

pub use self::{output::Output, parsing::arguments_from_field_attrs};

pub fn process_arguments<'a>(
    schema: &'a Schema<'a, Unvalidated>,
    literals: Vec<parsing::FieldArgument>,
    field: &crate::schema::types::Field<'a>,
    schema_module: syn::Path,
    variables_fields: Option<&syn::Path>,
    span: Span,
) -> Result<Output<'a>, Errors> {
    let analysed = analyse::analyse(schema, literals, field, variables_fields, span)?;

    Ok(Output {
        analysed,
        schema_module,
    })
}

#[cfg(test)]
mod tests;
