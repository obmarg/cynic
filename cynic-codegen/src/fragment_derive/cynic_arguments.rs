use proc_macro2::TokenStream;
use syn::{
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    Expr, Ident, Result, Token,
};

pub fn arguments_from_field_attrs(attrs: &Vec<syn::Attribute>) -> Result<Vec<FieldArgument>> {
    for attr in attrs {
        if attr.path.is_ident("cynic_arguments") {
            let parsed: CynicArguments = attr.parse_args()?;
            return Ok(parsed.arguments.into_iter().collect());
        }
    }
    Ok(vec![])
}

/// Implements syn::Parse to parse out arguments from the cynic_arguments
/// attribute.
#[derive(PartialEq, Debug)]
struct CynicArguments {
    arguments: Punctuated<FieldArgument, Token![,]>,
}

impl Parse for CynicArguments {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(CynicArguments {
            arguments: Punctuated::parse_terminated(input)?,
        })
    }
}

#[derive(PartialEq, Debug)]
pub enum ArgumentExpression {
    Literal(syn::Lit),
    FieldAccess(syn::ExprField),
}

impl std::convert::TryFrom<Expr> for ArgumentExpression {
    type Error = syn::Error;

    fn try_from(expr: Expr) -> Result<ArgumentExpression> {
        use syn::spanned::Spanned;

        match expr {
            Expr::Lit(expr) => Ok(ArgumentExpression::Literal(expr.lit)),
            Expr::Field(field_expr) => {
                if let Expr::Path(expr) = *field_expr.base.clone() {
                    if expr.path.is_ident("args") {
                        Ok(ArgumentExpression::FieldAccess(field_expr))
                    } else {
                        Err(syn::Error::new(
                            expr.span(),
                            "Query arguments are exposed on a struct named args",
                        ))
                    }
                } else {
                    Err(syn::Error::new(
                        field_expr.base.span(),
                        "Arguments should be of the form args.an_argument",
                    ))
                }
            }
            _ => Err(syn::Error::new(
                expr.span(),
                format!("Must be a literal or an expression of the form args.an_argument"),
            )),
        }
    }
}

impl quote::ToTokens for ArgumentExpression {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let result = match self {
            ArgumentExpression::Literal(expr) => quote! { #expr },
            ArgumentExpression::FieldAccess(expr) => quote! { #expr },
        };

        tokens.append_all(result);
    }
}

#[derive(PartialEq, Debug, Clone)]
pub struct FieldArgument {
    pub argument_name: Ident,
    //expr: ArgumentExpression,
    // TODO: Remove argument expression maybe?  seems too much work to be worth it just now.
    pub expr: syn::Expr,
}

impl Parse for FieldArgument {
    fn parse(input: ParseStream) -> Result<Self> {
        let argument_name = input.parse()?;
        input.parse::<Token![=]>()?;
        let expr: Expr = input.parse()?;

        Ok(FieldArgument {
            argument_name,
            expr, //expr.try_into()?,
        })
    }
}

impl quote::ToTokens for FieldArgument {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let argument_name = &self.argument_name;
        let expr = &self.expr;

        tokens.append_all(quote! {
            #argument_name: #expr
        });
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use syn::parse_quote;

    #[test]
    fn test_parsing_string_literal() {
        let parsed: CynicArguments = parse_quote! { x = "abcd" };

        let arguments = parsed.arguments.iter().collect::<Vec<_>>();

        assert_eq!(arguments.len(), 1);
        assert_eq!(arguments[0].argument_name.to_string(), "x".to_string());
    }

    #[test]
    fn test_parsing_multiple_arg_expressions() {
        let parsed: CynicArguments = parse_quote! { x = 1, y = args.test };

        let arguments = parsed.arguments.iter().collect::<Vec<_>>();

        assert_eq!(arguments.len(), 2);
        assert_eq!(arguments[0].argument_name.to_string(), "x".to_string());
        /*
        TODO: Re-instate

        assert_eq!(
            arguments[0].expr,
            ArgumentExpression::Literal(parse_quote! { 1 })
        );
        assert_eq!(arguments[1].argument_name.to_string(), "y".to_string());
        assert_matches!(arguments[1].expr, ArgumentExpression::FieldAccess(_));
        */
    }
}
