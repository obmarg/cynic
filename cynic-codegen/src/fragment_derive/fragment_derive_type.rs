use crate::schema::{
    markers::{FieldMarkerModule, MarkerIdent},
    types::{Field, Kind, Type},
    FieldName, SchemaError,
};

pub struct FragmentDeriveType<'a> {
    pub fields: Vec<Field<'a>>,
    pub name: &'a str,
    pub marker_ident: MarkerIdent<'a>,
    pub field_module: FieldMarkerModule<'a>,
}

impl<'a> FragmentDeriveType<'a> {
    pub fn field<N>(&self, name: &N) -> Option<&Field<'a>>
    where
        for<'b> FieldName<'b>: PartialEq<N>,
    {
        self.fields.iter().find(|field| field.name == *name)
    }
}

impl<'a> TryFrom<Type<'a>> for FragmentDeriveType<'a> {
    type Error = SchemaError;

    fn try_from(value: Type<'a>) -> Result<Self, Self::Error> {
        match value {
            Type::Interface(iface) => Ok(FragmentDeriveType {
                marker_ident: iface.marker_ident(),
                field_module: iface.field_module(),
                name: iface.name,
                fields: iface.fields,
            }),
            Type::Object(obj) => Ok(FragmentDeriveType {
                marker_ident: obj.marker_ident(),
                field_module: obj.field_module(),
                name: obj.name,
                fields: obj.fields,
            }),
            other => Err(SchemaError::unexpected_kind(other, Kind::ObjectOrInterface)),
        }
    }
}
