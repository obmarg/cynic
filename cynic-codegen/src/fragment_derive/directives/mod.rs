mod output;
mod parsing;

use proc_macro2::Span;
use syn::Ident;

use crate::{
    error::Errors,
    schema::{types::Directive, Schema, Unvalidated},
};

pub use self::{
    output::Output,
    parsing::{directives_from_field_attrs, FieldDirective},
};

use super::arguments::{analyse::AnalysedDirectiveArguments, ArgumentLiteral, FieldArgument};

pub struct AnalysedFieldDirective<'a> {
    directive: Directive<'a>,
    arguments: AnalysedDirectiveArguments<'a>,
}

pub fn process_directive<'a>(
    schema: &'a Schema<'a, Unvalidated>,
    directive: parsing::FieldDirective,
    variables_fields: Option<&syn::Path>,
    span: Span,
) -> Result<AnalysedFieldDirective<'a>, Errors> {
    match directive {
        FieldDirective::Skip(inner) => {
            let Some(directive) = schema.lookup_directive("skip")? else {
                return Err(syn::Error::new(span, "Unknown directive: skip").into());
            };

            let arguments = vec![FieldArgument {
                argument_name: Ident::new("if", span),
                value: super::arguments::FieldArgumentValue::Literal(match inner {
                    parsing::BooleanLiteral::Boolean(value) => {
                        ArgumentLiteral::Literal(syn::Lit::Bool(syn::LitBool { value, span }))
                    }
                    parsing::BooleanLiteral::Variable(var, span) => {
                        ArgumentLiteral::Variable(var, span)
                    }
                }),
            }];

            let arguments = super::arguments::analyse::analyse_directive_arguments(
                schema,
                arguments,
                &directive,
                variables_fields,
                span,
            )?;

            Ok(AnalysedFieldDirective {
                directive,
                arguments,
            })
        }
        FieldDirective::Include(inner) => {
            let Some(directive) = schema.lookup_directive("include")? else {
                return Err(syn::Error::new(span, "Unknown directive: include").into());
            };
            let arguments = vec![FieldArgument {
                argument_name: Ident::new("if", span),
                value: super::arguments::FieldArgumentValue::Literal(match inner {
                    parsing::BooleanLiteral::Boolean(value) => {
                        ArgumentLiteral::Literal(syn::Lit::Bool(syn::LitBool { value, span }))
                    }
                    parsing::BooleanLiteral::Variable(var, span) => {
                        ArgumentLiteral::Variable(var, span)
                    }
                }),
            }];

            let arguments = super::arguments::analyse::analyse_directive_arguments(
                schema,
                arguments,
                &directive,
                variables_fields,
                span,
            )?;

            Ok(AnalysedFieldDirective {
                directive,
                arguments,
            })
        }
        FieldDirective::Other { name, arguments } => {
            let Some(directive) = schema.lookup_directive(&name.to_string())? else {
                return Err(syn::Error::new(span, format!("Unknown directive: {name}")).into());
            };
            let arguments = super::arguments::analyse::analyse_directive_arguments(
                schema,
                arguments,
                &directive,
                variables_fields,
                span,
            )?;

            Ok(AnalysedFieldDirective {
                directive,
                arguments,
            })
        }
    }
}
