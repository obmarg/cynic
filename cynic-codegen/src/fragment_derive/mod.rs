use std::collections::{HashMap, HashSet};

use proc_macro2::{Span, TokenStream};
use syn::spanned::Spanned;

use crate::{
    load_schema, type_validation::check_types_are_compatible, Errors, FieldType, Ident, TypePath,
};

mod arguments;
mod schema_parsing;
mod type_ext;

pub(crate) mod input;

use arguments::{arguments_from_field_attrs, FieldArgument};
use schema_parsing::{Field, Object};
use type_ext::SynTypeExt;

pub use input::{FragmentDeriveField, FragmentDeriveInput};

use crate::suggestions::{format_guess, guess_field};
pub(crate) use schema_parsing::Schema;

pub fn fragment_derive(ast: &syn::DeriveInput) -> Result<TokenStream, syn::Error> {
    use darling::FromDeriveInput;

    match FragmentDeriveInput::from_derive_input(ast) {
        Ok(input) => load_schema(&*input.schema_path)
            .map_err(|e| Errors::from(e.into_syn_error(input.schema_path.span())))
            .map(Schema::from)
            .and_then(|schema| fragment_derive_impl(input, &schema))
            .or_else(|e| Ok(e.to_compile_errors())),
        Err(e) => Ok(e.write_errors()),
    }
}

pub fn fragment_derive_impl(
    input: FragmentDeriveInput,
    schema: &Schema,
) -> Result<TokenStream, Errors> {
    use quote::{quote, quote_spanned};

    input.validate()?;

    let schema_path = &input.schema_path;

    let object = schema
        .objects
        .get(&Ident::for_type(&input.graphql_type_name()))
        .ok_or_else(|| {
            syn::Error::new(
                input.graphql_type_span(),
                format!(
                    "Can't find {} in {}",
                    input.graphql_type_name(),
                    **schema_path
                ),
            )
        })?;

    let input_argument_struct = (&input.argument_struct).clone();
    let argument_struct = if let Some(arg_struct) = input_argument_struct {
        let span = arg_struct.span();

        let arg_struct_val: Ident = arg_struct.into();
        let argument_struct = quote_spanned! { span => #arg_struct_val };
        syn::parse2(argument_struct)?
    } else {
        syn::parse2(quote! { () })?
    };

    if let darling::ast::Data::Struct(fields) = &input.data {
        let graphql_name = &(input.graphql_type_name());
        let query_module = input.query_module;
        let fragment_impl = FragmentImpl::new_for(
            &fields,
            &input.ident,
            &object,
            Ident::new_spanned(&*query_module, query_module.span()).into(),
            graphql_name,
            argument_struct,
        )?;
        Ok(quote::quote! {
            #fragment_impl
        })
    } else {
        Err(syn::Error::new(
            Span::call_site(),
            "QueryFragment can only be derived from a struct".to_string(),
        )
        .into())
    }
}

/// Selector for a "field type" - i.e. a nullable/list/required type that
/// references some named schema type.
enum FieldTypeSelectorCall {
    Field(TypePath),
    Opt(Box<FieldTypeSelectorCall>),
    Vector(Box<FieldTypeSelectorCall>),
    Flatten(Box<FieldTypeSelectorCall>),
    Recurse(u8, Box<FieldTypeSelectorCall>, bool),
    Box(Box<FieldTypeSelectorCall>),
}

impl FieldTypeSelectorCall {
    fn for_field(
        field_type: &FieldType,
        field_constructor: TypePath,
        flatten: bool,
        recurse_limit: Option<u8>,
    ) -> FieldTypeSelectorCall {
        if flatten {
            FieldTypeSelectorCall::Flatten(Box::new(FieldTypeSelectorCall::for_field(
                field_type,
                field_constructor,
                false,
                None,
            )))
        } else if let Some(limit) = recurse_limit {
            let inner_selector = Box::new(FieldTypeSelectorCall::for_field(
                field_type,
                field_constructor,
                false,
                None,
            ));

            if field_type.is_list() {
                // List types can just recurse - no need for boxes
                FieldTypeSelectorCall::Recurse(limit, inner_selector, field_type.is_nullable())
            } else if field_type.is_nullable() {
                // Optional types need to be wrapped in Box to keep the rust compiler happy
                // i.e. `Box<Option<T>>`
                FieldTypeSelectorCall::Box(Box::new(FieldTypeSelectorCall::Recurse(
                    limit,
                    inner_selector,
                    field_type.is_nullable(),
                )))
            } else {
                // Required types need their inner types to be wrapped in box
                // i.e. `Option<Box<T>>`
                FieldTypeSelectorCall::Recurse(
                    limit,
                    Box::new(FieldTypeSelectorCall::Box(inner_selector)),
                    field_type.is_nullable(),
                )
            }
        } else if field_type.is_nullable() {
            FieldTypeSelectorCall::Opt(Box::new(FieldTypeSelectorCall::for_field(
                &field_type.clone().as_required(),
                field_constructor,
                false,
                None,
            )))
        } else if let FieldType::List(inner, _) = field_type {
            FieldTypeSelectorCall::Vector(Box::new(FieldTypeSelectorCall::for_field(
                &inner,
                field_constructor,
                false,
                None,
            )))
        } else {
            FieldTypeSelectorCall::Field(field_constructor)
        }
    }
}

impl FieldTypeSelectorCall {
    fn to_call(
        &self,
        required_arguments: &[FieldArgument],
        optional_arguments: &[FieldArgument],
        inner_selection_tokens: TokenStream,
    ) -> TokenStream {
        use quote::quote;

        match self {
            FieldTypeSelectorCall::Field(type_path) => {
                let required_arguments = required_arguments.iter().map(|arg| &arg.expr);
                let optional_arg_names = optional_arguments.iter().map(|arg| &arg.argument_name);
                let optional_arg_exprs = optional_arguments.iter().map(|arg| &arg.expr);

                quote! {
                    #type_path(
                        #(#required_arguments, )*
                    )
                    #(
                        .#optional_arg_names(#optional_arg_exprs)
                    )*
                    .select(#inner_selection_tokens)
                }
            }
            FieldTypeSelectorCall::Opt(inner) => inner.to_call(
                required_arguments,
                optional_arguments,
                inner_selection_tokens,
            ),
            FieldTypeSelectorCall::Vector(inner) => inner.to_call(
                required_arguments,
                optional_arguments,
                inner_selection_tokens,
            ),
            FieldTypeSelectorCall::Flatten(inner) => {
                let inner_call = inner.to_call(
                    required_arguments,
                    optional_arguments,
                    inner_selection_tokens,
                );

                quote! {
                    #inner_call.map(|item| {
                        use ::cynic::utils::FlattenInto;
                        item.flatten_into()
                    })
                }
            }
            FieldTypeSelectorCall::Recurse(limit, inner, field_nullable) => {
                let inner_call = inner.to_call(
                    required_arguments,
                    optional_arguments,
                    inner_selection_tokens,
                );

                let recurse_branch = if !field_nullable {
                    // If the field is required we need to wrap it in an option
                    quote! {
                        #inner_call.map(Some)
                    }
                } else {
                    // Otherwise, just let the fields option do the work.
                    inner_call
                };

                quote! {
                    if context.recurse_depth != Some(#limit) {
                        #recurse_branch
                    } else {
                        ::cynic::selection_set::succeed_using(|| None)
                    }
                }
            }
            FieldTypeSelectorCall::Box(inner) => {
                let inner_call = inner.to_call(
                    required_arguments,
                    optional_arguments,
                    inner_selection_tokens,
                );

                quote! {
                    #inner_call.map(Box::new)
                }
            }
        }
    }
}

/// The call style to use for a particular named type selector function
enum NamedTypeSelectorStyle {
    QueryFragment(syn::Type),
    Enum(syn::Type),
    Scalar,
}

struct FieldSelectorCall {
    selector_function: FieldTypeSelectorCall,
    style: NamedTypeSelectorStyle,
    required_arguments: Vec<FieldArgument>,
    optional_arguments: Vec<FieldArgument>,
    recurse_limit: Option<u8>,
    span: proc_macro2::Span,
}

impl quote::ToTokens for FieldSelectorCall {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, quote_spanned, TokenStreamExt};

        let span = self.span;

        let inner_selection_tokens = match (&self.style, self.recurse_limit) {
            (NamedTypeSelectorStyle::Scalar, _) => quote_spanned! {span => },
            (NamedTypeSelectorStyle::Enum(enum_type), _) => quote_spanned! {span =>
                #enum_type::select()
            },
            (NamedTypeSelectorStyle::QueryFragment(field_type), None) => quote_spanned! {span =>
                #field_type::fragment(context.with_args(FromArguments::from_arguments(args)))
            },
            (NamedTypeSelectorStyle::QueryFragment(field_type), Some(_)) => quote_spanned! {span =>
                #field_type::fragment(context.recurse().with_args(FromArguments::from_arguments(args)))
            },
        };

        let selector_function_call = &self.selector_function.to_call(
            &self.required_arguments,
            &self.optional_arguments,
            inner_selection_tokens,
        );

        tokens.append_all(quote! {
            #selector_function_call
        });
    }
}

struct ConstructorParameter {
    name: Ident,
    type_path: syn::Type,
}

impl quote::ToTokens for ConstructorParameter {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let name = &self.name;
        let type_path = &self.type_path;

        tokens.append_all(quote! {
            #name: #type_path
        })
    }
}

struct FragmentImpl {
    target_struct: Ident,
    fields: Vec<FieldSelectorCall>,
    selector_struct_path: TypePath,
    constructor_params: Vec<ConstructorParameter>,
    argument_struct: syn::Type,
    graphql_type_name: String,
}

impl FragmentImpl {
    fn new_for(
        fields: &darling::ast::Fields<FragmentDeriveField>,
        name: &syn::Ident,
        object: &Object,
        query_dsl_path: TypePath,
        graphql_type_name: &str,
        argument_struct: syn::Type,
    ) -> Result<Self, syn::Error> {
        let target_struct = Ident::new_spanned(&name.to_string(), name.span());
        let selector_struct_path =
            TypePath::concat(&[query_dsl_path, object.selector_struct.clone().into()]);
        let mut field_selectors = vec![];
        let mut constructor_params = vec![];

        if fields.style != darling::ast::Style::Struct {
            return Err(syn::Error::new(
                name.span(),
                "QueryFragment derive currently only supports named fields",
            ));
        }

        for field in &fields.fields {
            if let Some(macro_ident) = &field.ident {
                let field_name_span = macro_ident.span();
                let field_name = Ident::from(macro_ident.clone());

                constructor_params.push(ConstructorParameter {
                    name: field_name.clone(),
                    type_path: field.ty.clone(),
                });

                let arguments = arguments_from_field_attrs(&field.attrs)?;

                if let Some(gql_field) = object.fields.get(&field_name) {
                    check_types_are_compatible(
                        &gql_field.field_type,
                        &field.ty,
                        field.type_check_mode(),
                    )?;

                    let (required_arguments, optional_arguments) =
                        validate_and_group_args(arguments, gql_field, field_name_span)?;

                    field_selectors.push(FieldSelectorCall {
                        selector_function: FieldTypeSelectorCall::for_field(
                            &gql_field.field_type,
                            TypePath::concat(&[
                                selector_struct_path.clone(),
                                field_name.clone().into(),
                            ]),
                            field.flatten,
                            field.recurse.as_ref().map(|f| **f),
                        ),
                        style: if gql_field.field_type.contains_scalar() {
                            NamedTypeSelectorStyle::Scalar
                        } else if gql_field.field_type.contains_enum() {
                            NamedTypeSelectorStyle::Enum(field.ty.inner_type())
                        } else {
                            NamedTypeSelectorStyle::QueryFragment(field.ty.inner_type())
                        },
                        required_arguments,
                        optional_arguments,
                        recurse_limit: field.recurse.as_ref().map(|limit| **limit),
                        span: field.ty.span(),
                    })
                } else {
                    let candidates = object.fields.keys().map(|k| k.graphql_name());
                    let guss_value = guess_field(candidates, &(field_name.graphql_name()));
                    return Err(syn::Error::new(
                        field_name_span,
                        format!(
                            "Field {} does not exist on the GraphQL type {}.{}",
                            field_name.graphql_name(),
                            graphql_type_name,
                            format_guess(guss_value).as_str()
                        ),
                    ));
                }
            }
        }

        Ok(FragmentImpl {
            fields: field_selectors,
            target_struct,
            selector_struct_path,
            constructor_params,
            argument_struct,
            graphql_type_name: graphql_type_name.to_string(),
        })
    }
}

impl quote::ToTokens for FragmentImpl {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let argument_struct = &self.argument_struct;
        let target_struct = &self.target_struct;
        let selector_struct = &self.selector_struct_path;
        let fields = &self.fields;
        let constructor_params = &self.constructor_params;
        let graphql_type = proc_macro2::Literal::string(&self.graphql_type_name);
        let constructor_param_names = self
            .constructor_params
            .iter()
            .map(|p| &p.name)
            .collect::<Vec<_>>();

        let map_function = quote::format_ident!("map{}", fields.len());

        tokens.append_all(quote! {
            #[automatically_derived]
            impl ::cynic::QueryFragment for #target_struct {
                type SelectionSet = ::cynic::SelectionSet<'static, Self, #selector_struct>;
                type Arguments = #argument_struct;

                fn fragment(context: ::cynic::FragmentContext<Self::Arguments>) -> Self::SelectionSet {
                    use ::cynic::{QueryFragment, FromArguments, Enum};

                    let args = context.args;

                    let new = |#(#constructor_params),*| #target_struct {
                        #(#constructor_param_names),*
                    };

                    ::cynic::selection_set::#map_function(
                        new,
                        #(
                            #fields
                        ),*
                    )
                }

                fn graphql_type() -> String {
                    #graphql_type.to_string()
                }
            }
        })
    }
}

/// Validates the FieldArguments against the arguments defined on field
/// in the schema.  If everythings good, groups into required & optional
/// arguments in the correct order.
fn validate_and_group_args(
    arguments: Vec<FieldArgument>,
    field: &Field,
    missing_arg_span: Span,
) -> Result<(Vec<FieldArgument>, Vec<FieldArgument>), syn::Error> {
    let all_required: HashSet<Ident> = field
        .arguments
        .iter()
        .filter(|arg| arg.required)
        .map(|arg| arg.name.clone())
        .collect();

    let provided_names: HashSet<Ident> = arguments
        .iter()
        .map(|arg| arg.argument_name.clone().into())
        .collect();

    let missing_args: Vec<_> = all_required
        .difference(&provided_names)
        .map(|s| s.graphql_name())
        .collect();
    if !missing_args.is_empty() {
        let missing_args = missing_args.join(", ");

        return Err(syn::Error::new(
            missing_arg_span,
            format!("Missing arguments: {}", missing_args),
        ));
    }

    let all_arg_names: HashSet<Ident> =
        field.arguments.iter().map(|arg| arg.name.clone()).collect();

    let unknown_args: Vec<_> = provided_names
        .difference(&all_arg_names)
        .map(|s| s.graphql_name())
        .collect();

    if !unknown_args.is_empty() {
        let unknown_args = unknown_args.join(", ");

        return Err(syn::Error::new(
            missing_arg_span,
            format!("Unknown arguments: {}", unknown_args),
        ));
    }

    let provided_arguments: HashMap<Ident, _> = arguments
        .into_iter()
        .map(|a| (a.argument_name.clone().into(), a))
        .collect();

    let mut required = vec![];
    let mut optionals = vec![];
    for schema_arg in &field.arguments {
        if let Some(provided_arg) = provided_arguments.get(&schema_arg.name) {
            if schema_arg.required {
                required.push(provided_arg);
            } else {
                optionals.push(provided_arg);
            }
        }
    }

    Ok((
        required.into_iter().cloned().collect(),
        optionals.into_iter().cloned().collect(),
    ))
}
