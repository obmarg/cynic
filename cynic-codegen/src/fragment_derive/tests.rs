use std::{io::Write, process::Stdio};

use insta::assert_snapshot;
use rstest::rstest;
use syn::parse_quote;

use super::fragment_derive;

#[rstest(input => [
    parse_quote!(
        #[cynic(
            schema_path = "../cynic/tests/test-schema.graphql",
            schema_module = "schema",
            graphql_type = "Query"
        )]
        struct MyQuery {
            #[arguments(id: "1234")]
            post: Option<BlogPostOutput>,
            #[cynic(rename = "allPosts")]
            posts: Vec<BlogPostOutput>,
        }
    ),
    parse_quote!(
        #[cynic(
            schema_path = "../cynic/tests/test-schema.graphql",
            schema_module = "schema",
            graphql_type = "BlogPost"
        )]
        struct BlogPostOutput {
            has_metadata: Option<bool>,
            author: AuthorOutput,
        }
    ),
    parse_quote!(
        #[cynic(
            schema_path = "../cynic/tests/test-schema.graphql",
            schema_module = "schema",
            graphql_type = "Query"
        )]
        struct MyQuery {
            #[arguments(filters: {states: ["POSTED", "DRAFT"]})]
            filteredPosts: Vec<BlogPostOutput>,
        }
    ),
    parse_quote!(
        #[cynic(
            schema_path = "../cynic/tests/test-schema.graphql",
            schema_module = "schema",
            graphql_type = "Query",
            argument_struct = "AnArgumentStruct"
        )]
        struct MyQuery {
            #[arguments(filters: $filters)]
            filteredPosts: Vec<BlogPostOutput>,
        }
    ),
    parse_quote!(
        #[derive(cynic::QueryFragment, Debug)]
        #[cynic(
            schema_path = "../schemas/starwars.schema.graphql",
            query_module = "schema"
        )]
        struct Film {
            #[cynic(spread)]
            details: FilmDetails,
        }
    ),
    parse_quote!(
        #[derive(cynic::QueryFragment, Debug)]
        #[cynic(
            schema_path = "../schemas/starwars.schema.graphql",
            query_module = "schema"
        )]
        struct Film {
            #[cynic(flatten)]
            producers: String,
        }
    ),
    parse_quote!(
        #[cynic(
            schema_path = "../cynic/tests/test-schema.graphql",
            schema_module = "schema",
            graphql_type = "Query"
        )]
        struct MyQuery {
            #[arguments(filters: {states: ["POSTED", "POSTED"]})]
            filteredPosts: Vec<BlogPostOutput>,
        }
    ),
])]
fn snapshot_fragment_derive(input: syn::DeriveInput) {
    let tokens = fragment_derive(&input).unwrap();

    assert_snapshot!(format_code(format!("{}", tokens)));
}

fn format_code(input: String) -> String {
    let mut cmd = std::process::Command::new("rustfmt")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .expect("failed to execute rustfmt");

    write!(cmd.stdin.as_mut().unwrap(), "{}", input).unwrap();

    std::str::from_utf8(&cmd.wait_with_output().unwrap().stdout)
        .unwrap()
        .to_owned()
}
