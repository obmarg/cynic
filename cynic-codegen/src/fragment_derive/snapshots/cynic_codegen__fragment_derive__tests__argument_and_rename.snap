---
source: cynic-codegen/src/fragment_derive/tests.rs
expression: "format_code(format!(\"{}\", tokens))"
---
#[automatically_derived]
impl ::cynic::QueryFragment for MyQuery {
    type SchemaType = schema::Query;
    type VariablesFields = ();
    const TYPE: Option<&'static str> = Some("Query");
    fn query(
        mut builder: ::cynic::queries::SelectionBuilder<
            '_,
            Self::SchemaType,
            Self::VariablesFields,
        >,
    ) {
        #![allow(unused_mut)]
        let mut field_builder = builder . select_field :: < schema :: __fields :: Query :: post , < Option < BlogPostOutput > as :: cynic :: QueryFragment > :: SchemaType > () ;
        {
            field_builder
                .argument::<schema::__fields::Query::_post_arguments::id>()
                .literal("1234");
        }
        <Option<BlogPostOutput> as ::cynic::QueryFragment>::query(field_builder.select_children());
        let mut field_builder = builder . select_field :: < schema :: __fields :: Query :: allPosts , < Vec < BlogPostOutput > as :: cynic :: QueryFragment > :: SchemaType > () ;
        <Vec<BlogPostOutput> as ::cynic::QueryFragment>::query(field_builder.select_children());
    }
    fn name() -> Option<std::borrow::Cow<'static, str>> {
        Some(std::borrow::Cow::Borrowed("MyQuery"))
    }
}
#[automatically_derived]
impl<'de> ::cynic::serde::Deserialize<'de> for MyQuery {
    fn deserialize<__D>(deserializer: __D) -> Result<Self, __D::Error>
    where
        __D: ::cynic::serde::Deserializer<'de>,
    {
        #[derive(:: cynic :: serde :: Deserialize)]
        #[serde(field_identifier, crate = "::cynic::serde")]
        #[allow(non_camel_case_types)]
        enum __FragmentDeriveField {
            #[serde(rename = "post")]
            post,
            #[serde(rename = "allPosts")]
            posts,
            #[serde(other)]
            __Other,
        }
        struct Visitor<'de> {
            marker: ::core::marker::PhantomData<MyQuery>,
            lifetime: ::core::marker::PhantomData<&'de ()>,
        }
        impl<'de> ::cynic::serde::de::Visitor<'de> for Visitor<'de> {
            type Value = MyQuery;
            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                formatter.write_str("struct MyQuery")
            }
            fn visit_map<V>(self, mut __map: V) -> Result<Self::Value, V::Error>
            where
                V: ::cynic::serde::de::MapAccess<'de>,
            {
                let mut post = None;
                let mut posts = None;
                while let Some(__key) = __map.next_key()? {
                    match __key {
                        __FragmentDeriveField::post => {
                            if post.is_some() {
                                return Err(::cynic::serde::de::Error::duplicate_field("post"));
                            }
                            post = Some(__map.next_value()?);
                        }
                        __FragmentDeriveField::posts => {
                            if posts.is_some() {
                                return Err(::cynic::serde::de::Error::duplicate_field("allPosts"));
                            }
                            posts = Some(__map.next_value()?);
                        }
                        __FragmentDeriveField::__Other => {
                            __map.next_value::<::cynic::serde::de::IgnoredAny>()?;
                        }
                    }
                }
                let post = post.ok_or_else(|| ::cynic::serde::de::Error::missing_field("post"))?;
                let posts =
                    posts.ok_or_else(|| ::cynic::serde::de::Error::missing_field("allPosts"))?;
                Ok(MyQuery { post, posts })
            }
        }
        const FIELDS: &'static [&str] = &["post", "allPosts"];
        deserializer.deserialize_struct(
            "MyQuery",
            FIELDS,
            Visitor {
                marker: ::core::marker::PhantomData,
                lifetime: ::core::marker::PhantomData,
            },
        )
    }
}

